Userland/Libraries/LibWeb/Bindings/MainThreadVM.cpp

```diff
@@ -560,6 +565,45 @@ ErrorOr<void> initialize_main_thread_vm(HTML::EventLoop::Type type)
         HTML::fetch_single_imported_module_script(*module_map_realm, url.release_value(), *fetch_client, destination, fetch_options, *module_map_realm, fetch_referrer, module_request, perform_fetch, on_single_fetch_complete);
     };
 
+    // https://whatpr.org/html/9893/webappapis.html#hostinitializeshadowrealm(realm,-context,-o)
+    // 8.1.6.8 HostInitializeShadowRealm(realm, context, O)
+    s_main_thread_vm->host_initialize_shadow_realm = [](JS::Realm& realm, NonnullOwnPtr<JS::ExecutionContext> context, JS::ShadowRealm& object) -> JS::ThrowCompletionOr<void> {
+        // FIXME: 1. Set realm's is global prototype chain mutable to true.
+
+        // 2. Let globalObject be a new ShadowRealmGlobalScope object with realm.
+        auto global_object = HTML::ShadowRealmGlobalScope::create(realm);
+
+        // 3. Let settings be a new synthetic realm settings object that this algorithm will subsequently initialize.
+        auto settings = HTML::SyntheticRealmSettings {
+            // 4. Set settings's execution context to context.
+            .execution_context = move(context),
+
+            // 5. Set settings's principal realm to O's associated realm
+            .principal_realm = object.shape().realm(),
+
+            // 6. Set settings's underlying realm to realm.
+            .underlying_realm = realm,
+
+            // 7. Set settings's module map to a new module map, initially empty.
+            .module_map = realm.heap().allocate<HTML::ModuleMap>(realm),
+        };
+
+        // 8. Set realm.[[HostDefined]] to settings.
+        realm.set_host_defined(make<Bindings::SyntheticHostDefined>(move(settings)));
+
+        // 9. Set realm.[[GlobalObject]] to globalObject.
+        realm.set_global_object(global_object);
+
+        // 10. Set realm.[[GlobalEnv]] to NewGlobalEnvironment(globalObject, globalObject).
+        realm.set_global_environment(realm.vm().heap().allocate_without_realm<JS::GlobalEnvironment>(global_object, global_object));
+
+        // 11. Perform ? SetDefaultGlobalBindings(realm).
+        set_default_global_bindings(realm);
+
```

Comment:
we should call ` ShadowRealmGlobalScopeGlobalMixin::initialize(realm, *this);` here, as well as `add_shadow_realm_exposed_interfaces(*this)` (in a helper on the global scope/global object variable)

Reply:
speaking of which, did the spec folks decide what things should be exposed to shadow realms? We'll need to go around sprinkling annotations all over our idl files for the idl generator to pick them up

Reply:
Yes, spec folks have done so - those are all of the crashing IDL tests that we have currently. Here's a meta issue that shows what has been done: https://github.com/tc39/proposal-shadowrealm/issues/393

Reply:
Well, not crashing after these changes, but still doesn't work :D

Reply:
ah perfect, then it should be straightforward to add `ShadowRealm` as an option in `GenerateWindowOrWorkerInterfaces` ... which I guess should be renamed to `GenerateGlobalExposedInterfaces` or some such.

Reply:
ah! they added this new thing too. hm. https://whatpr.org/html/9893/webappapis.html#universalglobalscope

Reply:
Right, yeah that's why I've left those other FIXME's you mentioned in the initialize steps. I need to implement that universal global scope and reshuffling some implementations from windoworworkerglobal scope mixin to a new universalglobalscope mixin

Reply:
gotcha. well, given that, I think we're good to go on this part once CI is happy. Feel free to add more fixmes if you want though.